rest/admin-ui-ext/src/main/java/org/keycloak/admin/ui/rest/AvailableRoleMappingResource.java

```diff
@@ -176,8 +212,35 @@ public final List<ClientRole> listCompositeUserRoleMappings(@PathParam("id") Str
                     )
             )}
     )
-    public final List<ClientRole> listCompositeRoleMappings(@QueryParam("first") @DefaultValue("0") long first,
+    public final List<ClientRole> listAvailableRoleMappings(@PathParam("id") String id, @QueryParam("first") @DefaultValue("0") long first,
             @QueryParam("max") @DefaultValue("10") long max, @QueryParam("search") @DefaultValue("") String search) {
-        return this.mapping(o -> true, first, max, search);
+        if (this.auth.users().canManage() || !Profile.isFeatureEnabled(Profile.Feature.ADMIN_FINE_GRAINED_AUTHZ)) {
+            this.auth.users().requireManage();
+            return searchForClientRolesByExcludedIds(realm, search, first, max, Stream.of(id));
+        } else {
+            Set<String> roleIds = getRoleIdsWithPermissions(MAP_ROLE_COMPOSITE_SCOPE, MAP_ROLES_COMPOSITE_SCOPE);
+            roleIds.remove(id);
+            return searchForClientRolesByIds(realm, roleIds.stream(), search, first, max);
+        }
+    }
+
+    private Set<String> getRoleIdsWithPermissions(String roleResourceScope, String clientResourceScope) {
+        Set<String> roleIds = this.auth.roles().getRolesWithPermission(roleResourceScope);
+        Set<String> clientIds = this.auth.clients().getClientsWithPermission(clientResourceScope);
+        //TODO: this might be pretty inefficient, directly joining client IDs with role IDs in one query would be better
+        clientIds.stream().flatMap(cid -> realm.getClientById(cid).getRolesStream()).forEach(role -> roleIds.add(role.getId()));
+        return roleIds;
+    }
+
+    private List<ClientRole> searchForClientRolesByIds(RealmModel realm, Stream<String> includedIDs, String search, long first, long max) {
+        //TODO: not really sure about that casting of longs to ints here but I did not want to change the API sinature
```

Comment:
Feel free to change  `listAvailableRoleMappings()` to have Integers to be consistent with `UserResource` (see below) to avoid the casting. This would resolve this and the following TODO.

```
 @QueryParam("first") Integer firstResult, @QueryParam("max") Integer maxResults,
```